File: go-backend-interview-prep/07_io_files_networking/09_tcp_udp_basics/main.go
package main

import (
	"bufio"
	"fmt"
	"net"
	"time"
)

// echoServer accepts one TCP connection and echoes lines back.
func echoServer(ln net.Listener, ready chan<- struct{}) {
	ready <- struct{}{} // signal that server is listening

	conn, err := ln.Accept()
	if err != nil {
		return
	}
	defer conn.Close()

	scanner := bufio.NewScanner(conn)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "QUIT" {
			fmt.Fprintf(conn, "BYE\n")
			return
		}
		fmt.Fprintf(conn, "ECHO: %s\n", line)
	}
}

func main() {
	// --- TCP echo server + client ---
	fmt.Println("=== TCP Echo Server + Client ===")

	// Start server on random port
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		fmt.Println("listen error:", err)
		return
	}
	defer ln.Close()
	addr := ln.Addr().String()
	fmt.Println("  server listening on", addr)

	ready := make(chan struct{})
	go echoServer(ln, ready)
	<-ready // wait for server to be ready

	// Client connects
	conn, err := net.DialTimeout("tcp", addr, 2*time.Second)
	if err != nil {
		fmt.Println("  dial error:", err)
		return
	}
	defer conn.Close()
	fmt.Println("  client connected")

	// Send messages and read echoes
	messages := []string{"hello", "world", "Go networking", "QUIT"}
	reader := bufio.NewReader(conn)

	for _, msg := range messages {
		fmt.Fprintf(conn, "%s\n", msg)
		reply, err := reader.ReadString('\n')
		if err != nil {
			break
		}
		fmt.Printf("  sent: %q -> got: %s", msg, reply)
	}

	// --- UDP example (quick) ---
	fmt.Println("\n=== UDP (quick demo) ===")
	udpAddr, _ := net.ResolveUDPAddr("udp", "127.0.0.1:0")
	udpConn, err := net.ListenUDP("udp", udpAddr)
	if err != nil {
		fmt.Println("  udp listen error:", err)
		return
	}
	defer udpConn.Close()

	udpServerAddr := udpConn.LocalAddr().String()
	fmt.Println("  udp server on", udpServerAddr)

	// Client sends to server
	go func() {
		clientAddr, _ := net.ResolveUDPAddr("udp", udpServerAddr)
		client, _ := net.DialUDP("udp", nil, clientAddr)
		defer client.Close()
		client.Write([]byte("udp hello"))
	}()

	buf := make([]byte, 1024)
	udpConn.SetReadDeadline(time.Now().Add(2 * time.Second))
	n, remoteAddr, err := udpConn.ReadFromUDP(buf)
	if err != nil {
		fmt.Println("  udp read error:", err)
	} else {
		fmt.Printf("  received %d bytes from %s: %q\n", n, remoteAddr, buf[:n])
	}
}

File: go-backend-interview-prep/07_io_files_networking/09_tcp_udp_basics/main.py
"""TCP / UDP basics -- Python equivalent of the Go example.

Uses socket module for both TCP and UDP.
Server and client run in the same process using threads.
"""

import socket
import threading
import time


def echo_server(server_sock: socket.socket, ready: threading.Event) -> None:
    """Accept one connection and echo lines back."""
    ready.set()
    conn, addr = server_sock.accept()
    with conn:
        while True:
            data = conn.recv(1024)
            if not data:
                break
            line = data.decode().strip()
            if line == "QUIT":
                conn.sendall(b"BYE\n")
                return
            conn.sendall(f"ECHO: {line}\n".encode())


def main() -> None:
    # --- TCP echo server + client ---
    print("=== TCP Echo Server + Client ===")

    server_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server_sock.bind(("127.0.0.1", 0))
    server_sock.listen(1)
    addr = server_sock.getsockname()
    print(f"  server listening on {addr[0]}:{addr[1]}")

    ready = threading.Event()
    t = threading.Thread(target=echo_server, args=(server_sock, ready), daemon=True)
    t.start()
    ready.wait()

    # Client connects
    client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    client.settimeout(2.0)
    client.connect(addr)
    print("  client connected")

    messages = ["hello", "world", "Python networking", "QUIT"]
    for msg in messages:
        client.sendall(f"{msg}\n".encode())
        reply = client.recv(1024).decode()
        print(f"  sent: {msg!r} -> got: {reply}", end="")

    client.close()
    server_sock.close()

    # --- UDP example (quick) ---
    print("\n=== UDP (quick demo) ===")
    udp_server = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    udp_server.bind(("127.0.0.1", 0))
    udp_addr = udp_server.getsockname()
    print(f"  udp server on {udp_addr[0]}:{udp_addr[1]}")

    # Client sends to server in another thread
    def udp_client():
        time.sleep(0.05)
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        s.sendto(b"udp hello", udp_addr)
        s.close()

    threading.Thread(target=udp_client, daemon=True).start()

    udp_server.settimeout(2.0)
    try:
        data, remote = udp_server.recvfrom(1024)
        print(f"  received {len(data)} bytes from {remote}: {data.decode()!r}")
    except socket.timeout:
        print("  udp read timed out")

    udp_server.close()


if __name__ == "__main__":
    main()
